OpenSimRT/RealTime/src/RealTimeAnalysis.cpp

```diff
@@ -0,0 +1,243 @@
+                                                const SimTK::Vector& x,
+                                                const SimTK::Vector& xd,
+                                                const SimTK::Vector& xdd,
+                                                const int& nq) {
+    this->t = time;
+    this->q = x(0, nq);
+    this->qd = xd(0, nq);
+    this->qdd = xdd(0, nq);
+    this->externalWrenches = {};
+
+    auto wrenchSize = ExternalWrench::Input::size();
+    int wrenchCount = (x.size() - nq) / wrenchSize;
+
+    ExternalWrench::Input wrench;
+    for (int i = 0; i < wrenchCount; ++i) {
+        int offset = i * wrenchSize + nq;
+        wrench.fromVector(x(offset, offset + wrenchCount));
+        this->externalWrenches.push_back(wrench);
+    }
+}
+
+RealTimeAnalysis::RealTimeAnalysis(
+        const Model& otherModel, const RealTimeAnalysis::Parameters& parameters)
+        : model(*otherModel.clone()), parameters(parameters),
+          previousAcquisitionTime(-1.0), previousProcessingTime(-1.0),
+          notifyParentThread(false), terminationFlag(false) {
+    // filter
+    lowPassFilter = new LowPassSmoothFilter(parameters.filterParameters);
+
+    // ik
+    inverseKinematics = new InverseKinematics(
+            model, parameters.ikMarkerTasks, parameters.ikIMUTasks,
+            parameters.ikConstraintsWeight, parameters.ikAccuracy);
+
+    // id
+    inverseDynamics = new InverseDynamics(model, parameters.wrenchParameters);
+
+    // so
+    muscleOptimization = new MuscleOptimization(
+            model, parameters.muscleOptimizationParameters,
+            parameters.momentArmFunction);
+
+    // jr
+    jointReaction = new JointReaction(model, parameters.wrenchParameters);
+}
+
+bool RealTimeAnalysis::shouldTerminate() { return terminationFlag.load(); }
+
+void RealTimeAnalysis::shouldTerminate(bool flag) { terminationFlag = flag; }
+
+void RealTimeAnalysis::run() {
+    thread acquisitionThread(&RealTimeAnalysis::acquisition, this);
+    thread processingThread(&RealTimeAnalysis::processing, this);
+    acquisitionThread.detach();
+    processingThread.detach();
+}
+
+Vector RealTimeAnalysis::prepareUnfilteredData(
+        const Vector& q,
+        const vector<ExternalWrench::Input>& externalWrenches) const {
+    int m = q.size() + externalWrenches.size() * ExternalWrench::Input::size();
+    if (m == 0) { THROW_EXCEPTION("cannot convert from empty"); }
+
+    Vector v(m);
+    v(0, q.size()) = q;
+    for (int i = 0; i < externalWrenches.size(); ++i) {
+        auto w = externalWrenches[i].toVector();
+        v(q.size() + i * w.size(), w.size()) = w;
+    }
+    return v;
+}
+
+void RealTimeAnalysis::acquisition() {
+    try {
+        while (true) {
+            if (shouldTerminate()) THROW_EXCEPTION("Acquisition terminated.");
+
+            // get data
+            auto acquisitionData = parameters.dataAcquisitionFunction();
+            if (previousAcquisitionTime >= acquisitionData.IkFrame.t) {
```

Comment:
Maybe we should place things in a data buffer and notify this thread when we have a new frame.

Reply:
In this case, it is important that the acquisition thread has a higher FPS than the acquisition system to ensure that we consume everything with a constant dt for filtering.